src/Hazelcast.Net/Serialization/SerializationService.Register.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+
+namespace Hazelcast.Serialization
+{
+    internal partial class SerializationService
+    {
+        #region RegisterConstantSerializer
+
+        /// <summary>
+        /// Registers a constant serializer (by type and negative type-id).
+        /// </summary>
+        /// <param name="adapter">The serializer.</param>
+        /// <param name="type">The type.</param>
+        /// <exception cref="ArgumentNullException">The <paramref name="adapter"/> or the <paramref name="type"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentException">The <paramref name="adapter"/> has a positive type-id.</exception>
+        private void RegisterConstantSerializer(ISerializerAdapter adapter, Type type)
+        {
+            if (adapter == null) throw new ArgumentNullException(nameof(adapter));
+            if (adapter.TypeId > 0) throw new ArgumentException("Constant serializer type-id must be <=0.", nameof(adapter));
+            if (-adapter.TypeId >= ConstantSerializersCount) throw new ArgumentException($"Constant serializer type-id must be >-{ConstantSerializersCount}.", nameof(adapter));
+            if (type == null) throw new ArgumentNullException(nameof(type));
+
+            _ = _constantByType.TryAdd(type, adapter);
+            _constantById[-adapter.TypeId] = adapter; // type-id key is >0 because _constantById is an array
```

Comment:
I've looked but found nothing about negative `TypeId`. What is the aim of that?

Reply:
All the constant / default type serializers (for instance for a simple `int`) are registered with a negative id in order not to collide with custom serializers.